nx/test/nx/defn/evaluator_test.exs

```diff
@@ -478,4 +478,144 @@ defmodule Nx.Defn.EvaluatorTest do
       refute_received _
     end
   end
+
+  describe "cond cache" do
+    # The goal of those tests is to show that expressions inside cond are cached,
+    # regardless of evaluation order.
+    defn cond_cache_left(bool, a, b) do
+      res = hook(a + b, :example, &send_to_self({:hook, &1}))
+
+      cond =
+        if bool do
+          res
+        else
+          0
+        end
+
+      cond * res
+    end
+
+    test "on lhs" do
+      assert cond_cache_left(0, 1, 2) == Nx.tensor(0)
+      assert_received {:hook, _}
+      refute_received {:hook, _}
+
+      assert cond_cache_left(1, 1, 2) == Nx.tensor(9)
+      assert_received {:hook, _}
+      refute_received {:hook, _}
+    end
+
+    defn cond_cache_right(bool, a, b) do
+      res = hook(a + b, :example, &send_to_self({:hook, &1}))
+
+      cond =
+        if bool do
+          res
+        else
+          0
+        end
+
+      res * cond
+    end
+
+    test "on rhs" do
+      assert cond_cache_right(0, 1, 2) == Nx.tensor(0)
+      assert_received {:hook, _}
+      refute_received {:hook, _}
+
+      assert cond_cache_right(1, 1, 2) == Nx.tensor(9)
+      assert_received {:hook, _}
+      refute_received {:hook, _}
+    end
+
+    defn cond_cache_both(bool, a, b) do
+      res = hook(a + b, :example, &send_to_self({:hook, &1}))
+
+      left =
+        if bool do
+          res
+        else
+          -res
+        end
+
+      right =
+        if bool do
+          res * 2
+        else
+          res
+        end
+
+      left * right
+    end
+
+    test "on both" do
+      assert cond_cache_both(0, 4, 5) == Nx.tensor(-81)
+      assert_received {:hook, _}
+      refute_received {:hook, _}
+
+      assert cond_cache_both(1, 4, 5) == Nx.tensor(162)
+      assert_received {:hook, _}
+      refute_received {:hook, _}
+    end
+
+    defn cond_cache_map(state) do
+      state =
+        if state.iteration < 4 do
+          if state.iteration != 1 do
+            state
+          else
+            state
+          end
+        else
+          state
+        end
+
+      %{state | iteration: state.iteration + 1}
+    end
+
+    test "with nested map" do
+      assert cond_cache_map(%{iteration: 0}) ==
+               %{iteration: Nx.tensor(1)}
+
+      assert cond_cache_map(%{iteration: 0, abc: 1}) ==
+               %{iteration: Nx.tensor(1), abc: Nx.tensor(1)}
+
+      assert cond_cache_map(%{iteration: 0, xyz: 1}) ==
+               %{iteration: Nx.tensor(1), xyz: Nx.tensor(1)}
```

Comment:
Aren't these the same?

```suggestion
      assert cond_cache_map(%{iteration: 0, abc: 1}) ==
               %{iteration: Nx.tensor(1), abc: Nx.tensor(1)}
```

Reply:
No! they changed the map ordering which lead to different errors :D I will drop a comment.

Reply:
Oh wow :smile: